authenticatorbridge/src/main/java/com/bitwarden/authenticatorbridge/manager/AuthenticatorBridgeManagerImpl.kt

```diff
@@ -0,0 +1,195 @@
+                ?: false
+
+            if (!haveCorrectKey) {
+                // If we don't have the correct key, query for key:
+                symmetricKeyStorageProvider.symmetricKey =
+                    bridgeService.safeCall { symmetricEncryptionKeyData }
+            }
+
+            if (symmetricKeyStorageProvider.symmetricKey == null) {
+                // This means bridgeService returned a null key, which means we can make
+                // no valid operations. We should disconnect form the service and expose to the
+                // calling application that authenticator sync is not enabled.
+                mutableSharedAccountsStateFlow.value = AccountSyncState.SyncNotEnabled
+                unBindService()
+            }
+
+            // Register callback:
+            bridgeService.safeCall { registerBridgeServiceCallback(authenticatorBridgeCallback) }
+
+            // Sync data:
+            bridgeService.safeCall { syncAccounts() }
+        }
+
+        override fun onServiceDisconnected(name: ComponentName) {
+            bridgeService = AuthenticatorBridgeServiceSafetyWrapper(service = null)
+        }
+    }
+
+    override val accountSyncStateFlow: StateFlow<AccountSyncState> =
+        mutableSharedAccountsStateFlow.asStateFlow()
+
+    init {
+        // Listen for lifecycle events
+        processLifecycleOwner.lifecycle.addObserver(
+            object : DefaultLifecycleObserver {
+                override fun onStart(owner: LifecycleOwner) {
+                    bindService()
+                }
+
+                override fun onStop(owner: LifecycleOwner) {
+                    unBindService()
+                }
+            },
+        )
+    }
+
+    private fun bindService() {
+        val intent = Intent().apply {
+            component = ComponentName(
+                connectionType.toPackageName(),
+                AUTHENTICATOR_BRIDGE_SERVICE_CLASS,
+            )
+        }
+
+        isBound = try {
+            applicationContext.bindService(
+                intent,
+                bridgeServiceConnection,
+                Context.BIND_AUTO_CREATE,
+            )
+        } catch (e: SecurityException) {
+            false
+        }
+
+        if (!isBound) {
+            mutableSharedAccountsStateFlow.value = AccountSyncState.Error
+        }
+    }
+
+    private fun onAccountsSync(data: EncryptedSharedAccountData) {
+        // Received account sync update. Decrypt with local symmetric key and update StateFlow:
+        val decryptedAccounts = symmetricKeyStorageProvider.symmetricKey?.let {
+            val decryptedAccounts = data.decrypt(it).getOrNull()?.accounts ?: return@let null
+            AccountSyncState.Success(decryptedAccounts)
+        }
+        mutableSharedAccountsStateFlow.value = decryptedAccounts
+            ?.let { AccountSyncState.Success(it.accounts) }
+            ?: AccountSyncState.Error
+    }
+
+    private fun unBindService() {
+        if (!isBound) return
```

Comment:
Any reason to check this? Shouldn't we just all all this stuff no matter what?

Reply:
Calling `unbindService` if the service is not bound will throw an exception. Thus, the check.

We always call `unbindService` when we move to onStop. But there are other times we unbind, like if there are no accounts that have sync enabled, for example. So we have to be cautious here.

Reply:
OK. Yeah I guess this just goes into the discussion about what is safe and not safe to call here. So we can tie this up into that.